File: src/sub/singularvalues.c
/*
 * Compute singular values (eigenvalues) of symmetric matrix G
 * from Fortran SINGULARVALUES subroutine
 */

#include <stddef.h>
#include <stdio.h>
#include <stdbool.h> 

/* External global variables from vel_com.inc */
extern int nvar;
extern int nitt;
extern int ifixsolution;
extern bool single_turbo;
extern double g[];
extern double rht[];
extern double sv[];
extern double ale[];
extern double covs[];
extern double rs[];

/* External functions */
extern void svdsoluk(double A[4][4], double *RHt_ptr, int n, double scale,
                     double Xsol[4], double SV_ptr[], double *ale_val,
                     double COVs_ptr[], double Rs_ptr[]);
extern void alesubr(double SV_ptr[], int mode, double *ale_val);

void singularvalues(int i) {
    double A[4][4];
    double Xsol[4];
    int k, ii, jj, i1, j;
    
    /* Extract columns/rows from symmetric matrix G where k-th row or column appears */
    for (k = 0; k < nvar; k++) {
        ii = 0;
        jj = 0;
        /* Iterate over compressed symmetric storage (lower triangle + diagonal) */
        for (i1 = 0; i1 < nvar; i1++) {
            for (j = 0; j <= i1; j++) {
                /* Include element if it involves row/column k */
                if (i1 == k || j == k) {
                    A[k][ii] = g[jj];  /* 0-based access to g array */
                    ii++;
                }
                jj++;
            }
        }
    }
    
    /* Perform singular value decomposition */
    svdsoluk(A, rht, nvar, -1.0, Xsol, sv, &ale[i], covs, rs);
    
    /* Apply eigenvalue-based ALE if requested */
    if (ifixsolution == 1) {
        alesubr(sv, 3, &ale[i]);
    }
    if (ifixsolution == 9) {
        alesubr(sv, 1, &ale[i]);
    }
    
    /* Debug output if not in turbo mode */
    if (!single_turbo) {
        fprintf(stderr, "Singular values; iteration #%d\n", nitt);
        fprintf(stderr, "  ");
        for (jj = 0; jj < nvar; jj++) {
            fprintf(stderr, "%10.6f  ", sv[jj]);
        }
        fprintf(stderr, "\n");
        fprintf(stderr, "ALE = %g\n", ale[i]);
    }
}

File: src_fortran/sub/singularvalues.f
c
c
c
c
c
c
      subroutine SINGULARVALUES(i)
c
c     compute the singular values (in this case: eigenvalues) of the
c     symmetric matrix G.
c
      implicit none
      integer i
      include '../inc/vel_com.inc'
      integer k,ii,jj,i1,j
c
      real A(4,4), Xsol(4)
c
      do k=1,nvar
         ii=0
         jj=0
         do i1=1,nvar
            do j=1,i1
               jj=jj+1
               if(.NOT.(i1.ne.k.and.j.ne.k))then
                  ii=ii+1
                  A(k,ii)=g(jj)
               endif
            enddo
         enddo
      enddo
c
      call SVDSOLUK(A,RHt,nvar,-1.,Xsol,SV,ale(i),COVs,Rs)
c
      if(ifixsolution.eq.1) call ALESUBR(SV,3,ale(i))
      if(ifixsolution.eq.9) call ALESUBR(SV,1,ale(i))
c
      if(.not.single_turbo)then
         write(16,*)'Singular values; iteration #',nitt
         write(16,'(1x,4(2x,f10.6))') (sv(jj),jj=1,nvar)
         write(16,*)'ALE = ',ale(i)
      endif
c
ccc      write(16,*)'SVD-solution vector:'
ccc      write(16,'(1x,4(2x,f16.6))') (Xsol(jj),jj=1,nvar)
c
      return
      end ! of subr. singularvalues
